content/courses/advanced-ada/parts/data_types/records.rst

```diff
@@ -2918,20 +2928,684 @@ In this example, :ada:`Arr_Sub` and :ada:`Rec_Sub` are unconstrained subtypes.
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Device_Boost
+
+    package Devices is
+
+       type Device_State is
+         (Off, On);
+
+       type Device_Info is
+       record
+          V : Float;
+       end record;
+
+       type Device (State : Device_State;
+                    Boost : Boolean) is
+       record
+          case State is
+             when Off =>
+                null;
+             when On =>
+                Info : Device_Info;
+                case Boost is
+                   when False =>
+                      null;
+                   when True =>
+                      Factor : Float;
+                end case;
+          end case;
+       end record;
+
+    end Devices;
+
+In this version of the :ada:`Devices` package, we introduced a *boost button*
+as a discriminant (:ada:`Boost`) and an associated boost factor component
+(:ada:`Factor`) in the variant part.
+
+In the remaining of this section, we discuss a couple of details about variant
+records.
+
+.. admonition:: In the Ada Reference Manual
+
+    - :arm:`3.8.1 Variant Parts and Discrete Choices <3-8-1>`
+
+
+Discriminant type and value coverage
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The subtype of discriminants used in the variant part must be of a discrete
+type |mdash| it cannot be of an access or a floating-point type, for example.
+Also, all possible values of the subtype of each discriminant must be covered
+in the case statement of the variant part. For example, consider the following
+variant record:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Coverage
+    :class: ada-expect-compile-error
+
+    package Subtype_Coverage is
+
+       type Var_Rec (Value : Integer) is
+       record
+          case Value is
+             when 0 .. 100 =>
+                I : Integer;
+
+             --  ERROR: missing values!
+          end case;
+       end record;
+
+    end Subtype_Coverage;
+
+This package cannot be compiled because, in the variant part, we're only
+covering values for the :ada:`Value` discriminant in the range between 0 and
+100. To fix this compilation error, we have to cover all values instead. For
+example:
+
+.. code:: ada compile_button project=Courses.Advanced_Ada.Data_Types.Records.Variant_Parts.Coverage
+
+    package Subtype_Coverage is
+
+       type Var_Rec (Value : Integer) is
+       record
+          case Value is
+             when Integer'First .. -1 =>
+                null;
+             when 0 .. 100 =>
+                I : Integer;
+             when 101 .. Integer'Last =>
+                null;
+          end case;
+       end record;
+
+    end Subtype_Coverage;
+
+Of course, specifying all possible values can be difficult. As an alternative,
+we could simplify the case statement by just using :ada:`others` as a discrete
+choice that encompasses all values that haven't been specified elsewhere in the
```

Comment:
Maybe "elsewhere" should be "earlier" or "previously" because the "others choice" has to be last.

Reply:
Yes, good point, thanks!